### PR TITLE
Added json and binary logging feature to SPaT logger plugin

### PR DESCRIPTION
**Pull request by animeshb27**
# PR Details
## Related Issue
https://github.com/usdot-fhwa-OPS/V2X-Hub/issues/169

## Types of changes

- [ ] Defect fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [x] Breaking change (fix or feature that cause existing functionality to change)

## Checklist:

- [ ] I have added any new packages to the sonar-scanner.properties file
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
[V2XHUB Contributing Guide](https://github.com/usdot-fhwa-OPS/V2X-Hub/blob/develop/Contributing.md) 
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Review by arayama — COMMENTED** [review]
This looks good to me. Follows similar format as bsm logger so if tested, should be okay.

**Review by arayama — COMMENTED** [review]
Looks good to me.

File: src/v2i-hub/SPaTLoggerPlugin/src/SPaTLoggerPlugin.cpp [hunk 1]
```diff
@@ -1,236 +1,360 @@
...
+{
+
+	char *SpatOut;
+        cJSON *SpatRoot, *SpatMessageContent, *_SpatMessageContent;
+	
+	PLOG(logDEBUG)<<"HandleSPaTMessage";
+	auto spat = msg.get_j2735_data();
+	
+	std::stringstream direction_hex;
+        direction_hex<<"01";
```
Comment (dan-du-car) [inline src/v2i-hub/SPaTLoggerPlugin/src/SPaTLoggerPlugin.cpp:148]: This is value for  uint8_t   rxFrom; Could you use the unit8_t type to set the value, and then convert the value to hex string?

Comment (dan-du-car) [reply src/v2i-hub/SPaTLoggerPlugin/src/SPaTLoggerPlugin.cpp:148]:
```
typedef struct _SPaTMsgRecord {
    uint8_t   rxFrom; /* refer rxSource for values */
    intersection  curIntersection;
    uint32_t  utcTimeInSec;
    uint16_t  msec;
    int8_t    verificationStatus;
    uint16_t  length;
    /* payload of length size*/
} __attribute__((__packed__)) SPaTMsgRecord;

typedef struct _intersecion {
    int8_t intersectionStatus;
    int32_t intersectionId;
} __attribute__((__packed__)) intersection;
```
Where is this struct define?

File: src/v2i-hub/SPaTLoggerPlugin/src/SPaTLoggerPlugin.cpp [hunk 2]
```diff
@@ -1,236 +1,360 @@
...
+        else if (spatmillis16_int_hex.str()[3] == NULL){
+                spatmillis16_hex<<spatmillis16_int_hex.str()[2]<<"0"<<' '<<spatmillis16_int_hex.str()[0]<<spatmillis16_int_hex.str()[1];
+        }
+
+        else{
+                spatmillis16_hex<<spatmillis16_int_hex.str()[2]<<spatmillis16_int_hex.str()[3]<<' '<<spatmillis16_int_hex.str()[0]<<spatmillis16_int_hex.str()[1];
+        }
+
+	int spat_size;
+        if (routeableMsg.get_payload_str().length()%4 == 0){
```
Comment (dan-du-car) [inline src/v2i-hub/SPaTLoggerPlugin/src/SPaTLoggerPlugin.cpp:209]: Please add comments the purpose of setting the spat_size this way?

File: src/v2i-hub/SPaTLoggerPlugin/src/SPaTLoggerPlugin.cpp [hunk 3]
```diff
@@ -1,236 +1,360 @@
...
+        cJSON *SpatRoot, *SpatMessageContent, *_SpatMessageContent;
+	
+	PLOG(logDEBUG)<<"HandleSPaTMessage";
+	auto spat = msg.get_j2735_data();
+	
+	std::stringstream direction_hex;
+        direction_hex<<"01";
+
+        std::stringstream signStatus_hex;
+        signStatus_hex<<"00";
```
Comment (dan-du-car) [inline src/v2i-hub/SPaTLoggerPlugin/src/SPaTLoggerPlugin.cpp:151]: Is this supposed to be "verificationStatus" per the updated struct?

Comment (dan-du-car) [reply src/v2i-hub/SPaTLoggerPlugin/src/SPaTLoggerPlugin.cpp:151]: If yes, please change the variable name to verificationStatus. Or you may import the struct you defined and use the struct variable verificationStatus from there.

**Review by dan-du-car — APPROVED** [review]

**Comment by dan-du-car** [comment]
@arayama Please merge this if you are ok with it.

**Comment by dan-du-car** [comment]
close it as it is already addressed in this PR: https://github.com/usdot-fhwa-OPS/V2X-Hub/pull/203

**Pull request closed without merging**